### PR TITLE
[CLN] De-privatize commonly-used functions

### PR DESCRIPTION
**Pull request by jbrockmendel**
Also updated numpy_helper to not use things from numpy's deprecated C API.  This won't get rid of the warnings since cython still causes them, but it's still nice.

Not sure how to lint for this (or if we really want to), will see if google knows.

### CODE REVIEW

**Comment by jreback** [comment]
looks ok, something failing.

**Comment by jbrockmendel** [comment]
Messed up whitespace in rebasing, should be fixed now.

**Comment by jbrockmendel** [comment]
CI failure was a ResourceWarning.  Had to edit something to re-push, so de-privatized _ensure_categorical.

**Comment by jbrockmendel** [comment]
ping.  I'd like to get this out of the way before inevitable merge conflicts arise.

**Comment by jreback** [comment]
ahh i merged in a funny order, can you rebase

**Comment by jbrockmendel** [comment]
Done

File: pandas/core/common.py [hunk 1]
```diff
@@ -120,7 +120,7 @@ def is_bool_indexer(key):
     return False
 
 
-def _default_index(n):
+def default_index(n):
```
Comment (jreback) [inline pandas/core/common.py]: this entire module, is technically public, though we have a mix of public/private here anyway, so have to be careful about changing public functions here (obviously this is private -> public).

Comment (jreback) [reply pandas/core/common.py]: let' move this to ``pandas.index.base`` , ok to call it ``default_index``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/common.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/common.py]: done

**Review by jreback — APPROVED** [review]

**Pull request merged**